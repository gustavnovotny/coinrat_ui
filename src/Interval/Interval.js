// @flow
class Interval {
  since = null
  till = null

  constructor(since: ?Date = null, till: ?Date = null) {
    if (since !== null && till !== null && since > till) {
      throw Error('Since must be < than till')
    }

    this.since = since
    this.till = till
  }

  toIso = () => {
    return {
      since: this.since !== null ? this.since.toISOString() : null,
      till: this.till !== null ? this.till.toISOString() : null,
    }
  }

  withClosedFromRight = (date: Date) => {
    if (this.till === null) {
      const till = date > this.since ? date : this.since
      return new Interval(this.since, till)
    }
    return new Interval(this.since, this.till)
  }

  isEmpty = () => {
    return this.since.getTime() === this.till.getTime()
  }
}


export default Interval
