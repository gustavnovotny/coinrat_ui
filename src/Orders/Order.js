// @flow
import {extendObservable} from "mobx"

const DIRECTION_SELL = 'sell'
const DIRECTION_BUY = 'buy'

const STATUS_OPEN = 'open'
const STATUS_CLOSED = 'closed'
const STATUS_CANCELED = 'canceled'

const ORDER_TYPE_LIMIT = 'limit'
const ORDER_TYPE_MARKET = 'market'

type OrderDirectionType = DIRECTION_SELL | DIRECTION_BUY
type OrderType = ORDER_TYPE_LIMIT | ORDER_TYPE_MARKET
type OrderStatusType = STATUS_OPEN | STATUS_CLOSED | STATUS_CANCELED

class Order {
  orderId: string
  strategyRunId: string
  market: string
  direction: OrderDirectionType
  createdAt: Date
  pair: string
  type: OrderType
  quantity: string
  rate: string
  idOnMarket: string
  status: OrderStatusType
  closedAt: ?Date
  canceledAt: ?Date

  constructor(
    orderId: string,
    strategyRunId: string,
    market: string,
    direction: OrderDirectionType,
    createdAt: Date,
    pair: string,
    type: OrderType,
    quantity: string,
    rate: string,
    idOnMarket: string,
    status: OrderStatusType,
    closedAt: Date,
    canceledAt: Date
  ) {
    this.orderId = orderId
    this.strategyRunId = strategyRunId
    this.market = market
    this.direction = direction
    this.createdAt = createdAt
    this.pair = pair
    this.type = type
    this.quantity = quantity
    this.rate = rate
    this.idOnMarket = idOnMarket
    this.status = status
    this.closedAt = closedAt
    this.canceledAt = canceledAt
  }
}

class OrderDirectionAggregate {
  dateBucket: Date
  direction: OrderDirectionType
  countOpen: number
  countClosed: number
  countCanceled: number

  constructor(dateBucket: Date, direction: OrderDirectionType) {
    this.dateBucket = dateBucket
    this.direction = direction

    extendObservable(this, {
      countOpen: 0,
      countClosed: 0,
      countCanceled: 0,
    })
  }

  increment = (status: OrderStatusType) => {
    if (status === STATUS_OPEN) {
      this.countOpen++

    } else if (status === STATUS_CLOSED) {
      this.countClosed++

    } else if (status === STATUS_CANCELED) {
      this.countCanceled++

    } else {
      console.error('ERROR: unsupported order status: ', status)
    }
  }

  addAggregate = (aggregate: OrderDirectionAggregate) => {
    this.countOpen += aggregate.countOpen
    this.countClosed += aggregate.countClosed
    this.countCanceled += aggregate.countCanceled
  }

  maxValue = () => Math.max(this.countOpen, this.countClosed, this.countCanceled)
}

export type {
  OrderType,
  OrderDirectionType,
  OrderStatusType,
}

export {
  Order,
  OrderDirectionAggregate,

  ORDER_TYPE_LIMIT,
  ORDER_TYPE_MARKET,

  DIRECTION_SELL,
  DIRECTION_BUY,

  STATUS_OPEN,
  STATUS_CLOSED,
  STATUS_CANCELED,
}
