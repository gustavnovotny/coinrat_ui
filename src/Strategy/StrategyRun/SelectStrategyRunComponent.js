// @flow
import React from "react"
import {SelectComponent, SelectItemsType, SelectElement, SORTING_DESC} from "../../Form/Select/SelectComponent"

type Props = {
  availableStrategyRuns: SelectItemsType,
  onSelect: (strategyRunId: string) => void,
  defaultSelectedStrategyRun: string,
  isLoading: boolean
}

const SelectStrategyRunComponent = ({availableStrategyRuns, defaultSelectedStrategyRun, onSelect, isLoading}: Props) => {
  const item = availableStrategyRuns[defaultSelectedStrategyRun]
  return <SelectComponent
    label="Strategy Run"
    items={availableStrategyRuns}
    selectedItem={item || null}
    isLoading={isLoading}
    onChange={(strategyRun: SelectElement) => onSelect(strategyRun.key)}
    sort={SORTING_DESC}
  />
}

export default SelectStrategyRunComponent
