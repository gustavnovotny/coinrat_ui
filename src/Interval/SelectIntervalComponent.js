// @flow
import React, {Component} from "react"
import {DateRangeInput} from "@blueprintjs/datetime"
import {Label} from "@blueprintjs/core"
import Interval from "./Interval"
import appMainToaster from "../Toaster"
import {Box, Flex} from "reflexbox"

type Props = {
  defaultSelectedInterval: Interval,
  onChange: (interval: Interval) => void,
}

const lineStyles = {
  verticalAlign: 'middle',
  lineHeight: 30 + 'px',
  display: 'inline',
  marginLeft: 7 + 'px',
  minWidth: 120 + 'px',
}

class SelectIntervalComponent extends Component<Props> {
  props: Props

  handleChange = (data: Array<?Date>) => {
    let since = data[0]
    let till = data[1]
    since = Object.prototype.toString.call(since) === '[object Date]' && !isNaN(since.getTime()) ? since : null
    till = Object.prototype.toString.call(till) === '[object Date]' && !isNaN(till.getTime()) ? till : null

    if (since !== null && till !== null && since > till) {
      appMainToaster.show({message: "Since must be < than till.", className: 'pt-intent-danger'})
      return
    }

    this.props.onChange(new Interval(since, till))
  }

  render() {
    const interval = this.props.defaultSelectedInterval
    return <Flex>
      <Box w={120} style={{textAlign: 'right', ...lineStyles}}>
        <Label text='Interval:'/>
      </Box>
      <Box style={{textAlign: 'left', ...lineStyles}}>
        <DateRangeInput
          format="YYYY-MM-DD HH:mm:ss"
          value={[interval.since, interval.till]}
          onChange={this.handleChange}
          allowSingleDayRange={true}
        />
      </Box>
    </Flex>
  }
}

export default SelectIntervalComponent
