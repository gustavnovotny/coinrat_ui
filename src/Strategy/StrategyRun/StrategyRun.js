// @flow

import {Interval} from "../../Interval/Interval"

class StrategyRun {
  strategyRunId: string
  runAt: Date
  pair: string
  markets: string
  strategyName: string
  strategyConfiguration: string
  interval: Interval
  candleStorageName: string
  orderStorageName: string

  constructor(
    strategyRunId: string,
    runAt: Date,
    pair: string,
    markets: string,
    strategyName: string,
    strategyConfiguration: string,
    interval: Interval,
    candleStorageName: string,
    orderStorageName: string
  ) {
    if (interval.since === null) {
      throw new Error('Strategy run MUST have since time filled in interval.')
    }
    if (interval.till === null) {
      throw new Error('Strategy run MUST have till time filled in interval.')
    }
    this.strategyRunId = strategyRunId
    this.runAt = runAt
    this.pair = pair
    this.markets = markets
    this.strategyName = strategyName
    this.strategyConfiguration = strategyConfiguration
    this.interval = interval
    this.candleStorageName = candleStorageName
    this.orderStorageName = orderStorageName
  }

  getStrategyRunLengthInSeconds() {
    return this.interval.getAbsInSeconds()
  }

  getBaseCurrencyName(): string {
    return this.pair.split('_')[0]
  }

  getMarketCurrencyName(): string {
    return this.pair.split('_')[1]
  }
}

export {
  StrategyRun,
}
