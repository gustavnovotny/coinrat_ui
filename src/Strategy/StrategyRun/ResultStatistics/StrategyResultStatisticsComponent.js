// @flow
import React from "react"
import {StrategyRun} from "../StrategyRun"
import {Order} from "../../../Orders/Order"

type Props = {
  strategyRun: StrategyRun,
  orders: Array<Order>,
}

const StrategyResultStatisticsComponent = ({strategyRun, orders}: Props) => {
  const numberOfTrades = orders.length
  const tradesPerDay = numberOfTrades / (strategyRun.getStrategyRunLengthInSeconds() / (24 * 60 * 60))

  return <div style={{marginTop: 25 + 'px'}}>
    <h3>Statistics of strategy run</h3>
    <ul>
      <li>Number of trades: {numberOfTrades}</li>
      <li>Average trades per day: {tradesPerDay.toFixed(8)}</li>
    </ul>
  </div>
}

export default StrategyResultStatisticsComponent
