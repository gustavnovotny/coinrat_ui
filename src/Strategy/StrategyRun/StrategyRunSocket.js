// @flow
import {AppSocket} from "../../Sockets/socket"
import {
  SOCKET_EVENT_GET_STRATEGY_RUNS,
  SOCKET_EVENT_NEW_STRATEGY_RUN,
  SOCKET_EVENT_SUBSCRIBE,
  SOCKET_EVENT_UNSUBSCRIBE,
  SUBSCRIBED_EVENT_NEW_STRATEGY_RUN
} from "../../Sockets/SocketEvents"
import {StrategyRun} from "./StrategyRun"
import type {RawInterval} from "../../Interval/Interval"
import {deserialize_interval} from "../../Interval/Interval"

type RawStrategyRun = {
  strategy_run_id: string,
  run_at: string,
  pair: string,
  markets: string,
  strategy_name: string,
  strategy_configuration: string,
  interval: RawInterval,
  candle_storage_name: string,
  order_storage_name: string,
}

class StrategyRunSocket {
  socket: AppSocket

  constructor(socket: AppSocket) {
    this.socket = socket
  }

  registerNewStrategyRunEvent(processStrategyRuns: (strategyRun: Array<StrategyRun>) => void) {
    this.socket.socketio.on(SOCKET_EVENT_NEW_STRATEGY_RUN, (rawStrategyRun: RawStrategyRun) => {
      StrategyRunSocket.processStrategyRuns([rawStrategyRun], processStrategyRuns)
    })
  }

  loadStrategyRuns = (processStrategyRuns: (strategyRuns: Array<StrategyRun>) => void) => {
    this.socket.emit(SOCKET_EVENT_GET_STRATEGY_RUNS, {}, (status: String, rawStrategyRuns: Array<Object>) => {
      StrategyRunSocket.processStrategyRuns(rawStrategyRuns, processStrategyRuns)
      this.subscribeToStrategyRunsFeed()
    })
  }

  subscribeToStrategyRunsFeed() {
    this.socket.emit(
      SOCKET_EVENT_UNSUBSCRIBE,
      {event: SUBSCRIBED_EVENT_NEW_STRATEGY_RUN},
      () => {
        this.socket.emit(SOCKET_EVENT_SUBSCRIBE, {event: SUBSCRIBED_EVENT_NEW_STRATEGY_RUN})
      })
  }

  static processStrategyRuns(
    rawStrategyRuns: Array<RawStrategyRun>,
    processStrategyRuns: (strategyRuns: Array<StrategyRun>) => void
  ) {
    console.log('Received: STRATEGY_RUN', rawStrategyRuns.length)
    const strategyRuns = rawStrategyRuns.map(StrategyRunSocket.createStrategyRunFromRaw)
    processStrategyRuns(strategyRuns)
  }

  static createStrategyRunFromRaw = (rawStrategyRun: RawStrategyRun): StrategyRun => {
    return new StrategyRun(
      rawStrategyRun.strategy_run_id,
      new Date(Date.parse(rawStrategyRun.run_at)),
      rawStrategyRun.pair,
      rawStrategyRun.markets,
      rawStrategyRun.strategy_name,
      rawStrategyRun.strategy_configuration,
      deserialize_interval(rawStrategyRun.interval),
      rawStrategyRun.candle_storage_name,
      rawStrategyRun.order_storage_name
    )
  }
}

export {
  StrategyRunSocket,
}
