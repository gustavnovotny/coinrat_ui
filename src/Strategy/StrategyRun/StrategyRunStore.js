// @flow
import {StrategyRunSocket} from "./StrategyRunSocket"
import {action, extendObservable} from "mobx"
import {StrategyRun} from "./StrategyRun"

class StrategyRunStore {
  strategyRuns: Array<StrategyRun>

  constructor(strategyRunSocket: StrategyRunSocket) {
    this.strategyRunSocket = strategyRunSocket
    extendObservable(this, {
      strategyRuns: []
    })
  }

  reloadData = action((onSuccess: () => void): void => {
    this.strategyRunSocket.loadStrategyRuns((strategyRuns: Array<StrategyRun>) => {
      this.setStrategyRuns(strategyRuns)
      onSuccess()
    })
  })

  setStrategyRuns = action((strategyRuns: Array<StrategyRun>): void => {
    this.strategyRuns = strategyRuns
  })
}

export {
  StrategyRunStore,
}
