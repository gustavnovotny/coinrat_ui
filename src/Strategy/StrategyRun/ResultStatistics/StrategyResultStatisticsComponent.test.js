// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import StrategyResultStatisticsComponent from "./StrategyResultStatisticsComponent"
import {DIRECTION_SELL, Order, ORDER_TYPE_LIMIT, STATUS_CLOSED} from "../../../Orders/Order"
import {Interval} from "../../../Interval/Interval"
import {StrategyRun} from "../StrategyRun"

it('renders correctly', () => {
  const strategyRun = new StrategyRun(
    'this is ID',
    new Date(2018, 0, 1, 0, 0, 0),
    '',
    '',
    '',
    '',
    new Interval(new Date(2018, 0, 1, 0, 0, 0), new Date(2018, 0, 2, 0, 0, 0)),
    '',
    ''
  )
  const orders = [new Order(
    'ABCDEFGH',
    'yoloolooolololololooo',
    'foo_market',
    DIRECTION_SELL,
    new Date(2018, 0, 1, 0, 0, 0),
    'USD_BTC',
    ORDER_TYPE_LIMIT,
    '13',
    '8000',
    '',
    STATUS_CLOSED,
    new Date(2018, 0, 1, 0, 0, 0),
    null
  )]
  const tree = renderer
    .create(
      <StrategyResultStatisticsComponent orders={orders} strategyRun={strategyRun}/>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
