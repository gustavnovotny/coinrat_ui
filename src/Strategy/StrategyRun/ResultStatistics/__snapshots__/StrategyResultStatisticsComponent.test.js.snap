// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "marginTop": "25px",
    }
  }
>
  <h3>
    Statistics of strategy run
  </h3>
  <ul>
    <li>
      Number of trades: 
      1
    </li>
    <li>
      Average trades per day: 
      1.00000000
    </li>
  </ul>
</div>
`;
